src/main/java/org/reactome/addlinks/dataretrieval/pharos/PharosDataRetriever.java

```diff
@@ -0,0 +1,159 @@
+				int retries = this.numRetries;
+				boolean done = retries + 1 <= 0;
+				while (!done)
+				{
+					try (CloseableHttpClient client = HttpClients.createDefault();
+						CloseableHttpResponse response = client.execute(post, context); )
+					{
+						int statusCode = response.getStatusLine().getStatusCode();
+						// If status code was not 200, we should print something so that the users know that an unexpected response was received.
+						if (statusCode != HttpStatus.SC_OK)
+						{
+							if (String.valueOf(statusCode).startsWith("4") || String.valueOf(statusCode).startsWith("5"))
+							{
+								logger.error("Response code was 4xx/5xx: {}, Status line is: {}", statusCode, response.getStatusLine());
+							}
+							else
+							{
+								logger.warn("Response was not \"200\". It was: {}", response.getStatusLine());
+							}
+						}
+						else
+						{
+							// result could be flattened with jq using the jq expression: ".data.targets.targets[].uniprot"
+							String jsonResponse = EntityUtils.toString(response.getEntity());
+							JSONObject jsonObj = new JSONObject(jsonResponse);
+
+							int numProcessed = processJSONArray(fw, jsonObj);
+							moreRecords = numProcessed >= 1;
+							logger.info("Completed request #{}", numRequests);
+							numRequests++;
+						}
+						done = true;
+					}
+					catch (ConnectTimeoutException e)
+					{
+						// we will only be retrying the connection timeouts, defined as the time required to establish a connection.
+						// we will not handle socket timeouts (inactivity that occurs after the connection has been established).
+						// we will not handle connection manager timeouts (time waiting for connection manager or connection pool).
+						e.printStackTrace();
+						logger.info("Failed due to ConnectTimeout, but will retry {} more time(s).", retries);
+						retries--;
+						done = retries + 1 <= 0;
+						if (done)
+						{
+							throw new Exception("Connection timed out. Number of retries ("+this.numRetries+") exceeded. No further attempts will be made.", e);
+						}
+					}
+					catch (HttpHostConnectException e)
+					{
+						logger.error("Could not connect to host {} !",post.getURI().getHost());
+						e.printStackTrace();
+						throw e;
+					}
+					catch (IOException e) {
+						logger.error("Exception caught: {}",e.getMessage());
+						throw e;
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Implementors must return a specific query to be executed by this class' doHttpDownload
+	 * @return
+	 */
+	protected abstract String getQuery();
+
+	/**
+	 * Implementors must process an array of JSON objects.
+	 * @param fw A FileWriter - this can be used to write data to the output file.
+	 * @param jsonObj - the JSON object to process.
+	 * @return The number of items processed.
+	 * @throws IOException
+	 */
+	protected abstract int processJSONArray(FileWriter fw, JSONObject jsonObj) throws IOException;
```

Comment:
Could the FileWriter parameter be `fileWriter` instead of `fw` just so it's clear what it is when in the body of the code without the type declaration beside it?